### PR TITLE
WIP: Run framework agnostic tests

### PR DESCRIPTION
**Pull request by panpann**
Hello,

# context
a couple weeks ago, I created a branch to add [quart](https://github.com/pgjones/quart) support to connexion. I managed to implement the app and api but I struggled when it came time to test my work.
I wasn't familiar at all with the connexion codebase and didn't feel able to add tests.  
Also this was frustrating because Flask and Quart follow almost the same API (except async for quart). I just wanted to use all tests that request the tests API via `test_client()`.  

I like your idea of API first principle, it makes APIs consistent by taking the time to think. And to write an openapi before starting coding helps to normalize data sent and returned by the API. It is easy to implement well documented API contracts and keep the code clean and business focused.
 
In the future, I would like to have the possibility to choose any python framework and to use connexion on it.
I think that to test the same way each implementation could help contributors to add more frameworks to connexion, if the contributor only has to implement the interface and see a concrete error if its code is wrong. It would also ensure that whatever server I'm using, the API is consistent and responds the same way.

# proposed changes
I've started working on it,
1. I added an API `AbstractApp.test_client()` which returns an `AbstractClient`, each web server has to implement 1 **synchronous** method to request the API and returns an enhanced version of ConnexionResponse (just added helpers like response.json, response.text).
For now the API to send request via this AbstractClient respects the client's flask API. 
2. I've changed all tests that were using app.app.test_client() to app.test_client() and all tests passed.
3. I created the test_client for the AioHttpApp
4. I am now trying to pass all tests with AioHttpApp, and I encountered [this issue](https://github.com/zalando/connexion/issues/578). Are you okay if I'm making changes to respect the same return API that flask ? 

Are you okay with changes I'm proposing in this pull request ?

The work currently won't pass the CI, I'm creating this pull request to share my idea with you and have your opinion on it

### CODE REVIEW

File: connexion/lifecycle.py [hunk 1]
```diff
@@ -34,8 +38,39 @@ def __init__(self,
                  content_type=None,
                  body=None,
                  headers=None):
+        if not isinstance(status_code, int) or not (100 <= status_code <= 505):
+            raise ValueError("{} is not a valid status code".format(status_code))
```
Comment (dtkav) [inline connexion/lifecycle.py]: Should connexion really be policing which status codes the user is allowed to send?

**Comment by dtkav** [comment]
Hey @panpann - I really like the idea of patching up the differences in the API between the different frameworks. I definitely applaud your effort to fix this. I'm also excited about the idea of being compatible with more frameworks like quart.

I started reading through your PR, and it's quite long and included many orthogonal changes.

Can you think about how you might break it up into chunks that all make sense on their own?
For example, large scale whitespace changes should be done separately from changes in functionality.

Some things that could make sense as standalone fixes/improvements:
* `app.app.test_client()` -> `app.test_client()`
* additional parameterized tests
* remove werkzeug workaround in `test_post_wrong_content_type`
* max character length whitespace fixes

I'd love to review any of these individually, and continue to discuss the broader goal of cleaner abstractions, and supporting more frameworks.

Sorry I wasn't able to get back to you sooner. I was on holiday. Great work!

**Comment by Jyhess** [comment]
I like the idea to support tuple as return values for AioHttp handler, like for flask.
I did the same job in https://github.com/Jyhess/connexion/tree/return_tuple_for_aiohttp and discover your pull request just before submitting mine.
Is it make sense for me to continue on this way as a first step of this bigger work ?

**Comment by dtkav** [comment]
Hey @panpann and @Jyhess I'm very keep to implement these changes.
As I said, it's ideal to break them up into smaller and more targeted diffs. If you like I can probably help with that at the end of the week, but if you have something ready I'd be happy to review it.

**Comment by Jyhess** [comment]
@panpann @dtkav I let you review https://github.com/zalando/connexion/pull/849

**Comment by panpann** [comment]
Hi @dtkav

I wasn't available last few days, I know there is a lot of changes in this PR, this is because I'm debugging one test file in **tests/api/** folder after another.
One way to ease the review could be to create small PRs when a functionnality emerge from this mess :)

Do you have another solution ?

**Comment by dtkav** [comment]
That sounds great. @Jyhess has already been pulling off smaller chunks too, so looks like there's some good collaboration! I'm excited about the direction this is going.

**Comment by panpann** [comment]
OK I pushed my last local changes as a reference.

My first step will be to create the `App.test_client` for aiohttp and flask.
This is what I've done in this branch but then I ran all tests with flask and aiohttp apps and errors happened..

I will find a way to explicitly set a test as running in FlaskApp or AioHttpApp or both, something like
```python
@pytest.mark.app("FlaskApp")
def test_specific_for_flask():
    pass
```

This will allow to incrementally run both Flaskapp and AioHttpApp on specific tests covered by thoses small PRs

**Comment by cognifloyd** [comment]
After quart support gets in, I'd love to see sanic join the connexion crew as well :) this is awesome stuff. Is #849 the only break out PR so far?

**Comment by cognifloyd** [comment]
This will need to be rebased, but probably after #849. Some of the commits will drop out as they are fixed or implemented elsewhere.

58e3fbb8e3d74b7b48567e2861f8f94e6c4ddc56 is implemented in #849 (ready)
8eda5e3156f5764f79e3d7b0d0347e293051732c was implemented in #952 (merged)
67e7dacdfbb5cf6ae4e6b10121d6afc9abe2bebb looks like it is no longer needed.

These are moving things to operation validation - Why? does that make sense?
b8d979fdad946f540f01830621a80e576c185af8 I'm not sure how to reimplement this with the new _jsonify_data in abstract, or even if we should.
e488a6f78486633244ec499839943327f8655e45 This bit moved to abstract api.

Parts of these are candidates for dedicated coding style PRs:
7c0cdd3472406c4979f3a242ebfe3a997cf3ec99 ws: tests.test_json_validateion
9eaadafaeb829b86a912d348d36862270274d931 ws: tests.fakeapi.hello
1d40dda7072754046eed16de0daf7953532c7239 imports

It looks like the other commits fall into three general categories:
1. Generalize ConnexionRequest ConnexionResponse to make them useful for testing
2. Generalize a TestClient
3. Adjust the tests to use the generic test client and parametrize them to use both implementations

**Comment by ShadowJonathan** [comment]
Any chance this pull request can be pushed forward?

**Comment by RobbeSneyders** [comment]
Closing as this is based on outdated code. Since we're splitting a lot of functionality into framework-agnostic middleware, this will become a lot easier. Would still be interested to review a PR for a connexion test client.

**Pull request closed without merging**